server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java

```diff
@@ -0,0 +1,523 @@
+            assert metaData.getIndexUUID().equals(IndexMetaData.INDEX_UUID_NA_VALUE) == false;
+            document.add(new StringField(INDEX_UUID_FIELD_NAME, metaData.getIndexUUID(), Field.Store.NO));
+            indexWriter.addDocument(document);
+        }
+
+        private Document buildDocument(String typeName, ToXContent metaData) throws IOException {
+            final Document document = new Document();
+            document.add(new StringField(TYPE_FIELD_NAME, typeName, Field.Store.NO));
+
+            final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+            try (XContentBuilder xContentBuilder = XContentFactory.contentBuilder(XContentType.SMILE, outputStream)) {
+                xContentBuilder.startObject();
+                metaData.toXContent(xContentBuilder, FORMAT_PARAMS);
+                xContentBuilder.endObject();
+            }
+            document.add(new StoredField(DATA_FIELD_NAME, new BytesRef(outputStream.toByteArray())));
+            return document;
+        }
+    }
+
+    static class LucenePersistedState implements CoordinationState.PersistedState, Closeable {
+
+        private long currentTerm;
+        private ClusterState lastAcceptedState;
+        private final List<MetaDataIndex> metaDataIndices;
+        private final String nodeId;
+
+        LucenePersistedState(String nodeId, List<MetaDataIndex> metaDataIndices, long currentTerm, ClusterState lastAcceptedState) {
+            this.currentTerm = currentTerm;
+            this.lastAcceptedState = lastAcceptedState;
+            this.metaDataIndices = metaDataIndices;
+            this.nodeId = nodeId;
+        }
+
+        @Override
+        public long getCurrentTerm() {
+            return currentTerm;
+        }
+
+        @Override
+        public ClusterState getLastAcceptedState() {
+            return lastAcceptedState;
+        }
+
+        void persistInitialState() throws IOException {
+            for (MetaDataIndex metaDataIndex : metaDataIndices) {
+                metaDataIndex.persistInitialState(nodeId, currentTerm, lastAcceptedState);
+            }
+        }
+
+        @Override
+        public void setCurrentTerm(long currentTerm) {
+            persist(currentTerm, lastAcceptedState.version());
+            this.currentTerm = currentTerm;
+        }
+
+        @Override
+        public void setLastAcceptedState(ClusterState clusterState) {
+            try {
+                if (clusterState.term() != lastAcceptedState.term()) {
+                    assert clusterState.term() > lastAcceptedState.term() : clusterState.term() + " vs " + lastAcceptedState.term();
+                    // In a new currentTerm, we cannot compare the persisted metadata's lastAcceptedVersion to those in the new state, so
+                    // it's simplest just to write everything again.
+                    for (MetaDataIndex metaDataIndex : metaDataIndices) {
```

Comment:
Now we're serializing everything over and over for each separate data path? Shouldn't we at least create the `Document` only once and then loop over all writers? Or better yet, why not continue doing what we did so far and simply write to one data path and then repo to the others after writing? (saving us from having merges running for each path and doing all the work of managing the index in general multiple times?)

Reply:
Ok, I adjusted things in 2bf23b68232780bd23df687a19fff0f6e84eb80c to loop over docs first and then indices so we can (a) re-use the same `Document` for each index and then (b) discard it once written.

I would rather not do the write-and-copy dance here. The reason for having multiple data paths on a (dedicated) master node would be for protection against flaky storage, but write-and-copy risks propagating a write-time corruption on the first path across all paths.

I did briefly investigate a `Directory` implementation that replicates writes across multiple paths, allowing us to use a single `IndexWriter` instead, but it seemed tricky to implement things like `rename()` in a suitably atomic fashion.

Reply:
@DaveCTurner 

> The reason for having multiple data paths on a (dedicated) master node would be for protection against flaky storage, but write-and-copy risks propagating a write-time corruption on the first path across all paths.

We could just do a manual copy like we do for snapshots by copying all the files via a read from something analogous to `org.elasticsearch.index.store.Store#openVerifyingInput` which checks each file's checksum to avoid copying corrupted bytes? That would probably still be a ton faster than going through all the indexing steps for each directory separately.
Technically, you'd only have to do that manual copying once and could then do a real copy for subsequent paths if there's more than 2 because you verified consistency in the first copy.

Reply:
I do not like the thought of _three_ different ways of writing data depending on whether you have 1, 2 or ≥3 data paths. Rarely-exercised code is a breeding ground for bugs and I would rather not introduce it without a compelling reason to do so. We do not today have very strong testing around things like `fsync()` failures and I fear we could be tripped up by some of the stranger production environments out there.

I have added an item to the meta-issue to revisit this with some experiments at a later date. I expect the actual indexing to be super-cheap compared with the unavoidable cost of fsyncing, but we shall see...

Reply:
> I expect the actual indexing to be super-cheap compared with the unavoidable cost of fsyncing

++ (at least in those cases where `fsync` creates pain today :))